replace recursive get children call by single sql query

veloutin opened the pull request:
This replaces the recursive account children call by a single query.

I have an issue with an environment failing tests because of a badly ordered return of this function, and I couldn't find the cause, so I made a replacement.

Commit:
replace recursive get children call by single sql query

Commit:
fix: remove superfluous () around sql IN %s

Commit:
call _parent_store_compute if parents are not set

clonedagain commented:
Was it proposed upstream, at least in master? Or is it a backport of a later version?

hbrunn commented on addons/account/account.py:
```diff
@@ -269,15 +269,28 @@ def search(self, cr, uid, args, offset=0, limit=None, order=None,
-        return ids2 + ids3
+        res = self._query_children_and_consol(cr, uid, ids, context=context)
+        if ids and not res:
+            self._parent_store_compute(cr)
```
should we really do this? If we encounter this situation, something messed up `parent_{left,right}`, and we shouldn't fix it secretly I think. Also the function name implies reading, not fixing up parent/child relations.

veloutin replied:
I've had this situation bite me in unit tests, where the whole table had NULLS in parent fields. I don't know what else to do in the case where the relations aren't set.
I'd say that forcing the tree to be computed is fair game if you want to be able to query it. This is similar to calling functions for computed fields.

hbrunn replied:
@veloutin it's still fixing symptoms. A stored function field would be the right thing to compare, and there you also get null if it wasn't ever computed.

If it's only about unittests, call `_parent_store_compute` there.

This btw will trigger `_parent_store_compute` if you call it on accounts that happen not to have children, we don't want that in any case.

veloutin replied:
I have to agree about fixing symptoms :8ball: 

About accounts with no children triggering this, normally _get_children_and_consol returns the id of the parent, so when it doesn't return itself.

I've removed the _parent_store_compute. Besides, the CTE query goes around the parent_foo problem, thankfully.

hbrunn commented on addons/account/account.py:
```diff
@@ -269,15 +269,28 @@ def search(self, cr, uid, args, offset=0, limit=None, order=None,
+                ON parent.parent_left <= child.parent_left
+               AND parent.parent_right >= child.parent_right
+            WHERE parent.id IN %s
+            ORDER by COALESCE(child.level, 0), child.parent_left
```
up to now, you mimic the `childof` operator, but you miss including the consolidated children.

A left join on `account_account` via `account_account_consol_rel` should suffice.

veloutin replied:
Well... Either I use a recursive (iterative) SQL query (PG 9+), or this whole change is now pointless since we still have to loop on consolidated children. Unless someone has a better idea, I suppose this should be closed.

Left join on account_account will not get children of consolidated children.

hbrunn replied:
right, not that simple.

Judjing from https://github.com/odoo/odoo/search?utf8=%E2%9C%93&q=%22with+recursive%22&type=Code, CTEs should be okay.

veloutin replied:
Good news! I'll test and submit my CTE version.

veloutin commented:
@clonedagain It hasn't been submitted in upstream yet, but it will once reviewing and fixing is done here.

Commit:
cover consolidated children with a CTE query

Commit:
do not check and call parent_store_compute

bwrsandman commented:
Can this be reused in other contexts?

For example, if you moved that query into orm and factorized the table and fields into parameters. Could it be useful to replace other queries?

veloutin commented:
@bwrsandman This can be applied to all tables with a parent_id relationship and optionally a m2m relationship as well. The where condition in the second part of the query would need to be adjusted according to the tables in play.

So, yes it can be factorized. I don't know of any specific location where this code is needed, though.

Commit:
Fix parents being returned before their children

In cases where you have A > B > C, if you ask for children of
A and C you could end up with A returned before C, this sorts
data by depth and returns the deeper data first.

Commit:
remove extra comma in sql query

Commit:
return parents before children, as it will get reversed

veloutin commented:
This probably needs a few tests to make sure that parents are always returned before their children, even in cases like A > B > C, and you call _get_children_and_consol([C, A]). This is important because of the __compute function for balances that reverts the list and then sums up any children.

If _get_children_and_consol([C, A]) gives you C, A, B, __compute will make a key error when computing B because C isn't computed yet.

dufresnedavid commented on addons/account/account.py:
```diff
@@ -269,15 +269,33 @@ def search(self, cr, uid, args, offset=0, limit=None, order=None,
+                        WHERE a.parent_id = ac.id OR a.id IN
+                                (SELECT parent_id
+                                 FROM account_account_consol_rel a
+                                 WHERE a.child_id = ac.id)
```
@veloutin This part of your query is a bit confusing. You use the variable "a" to refer to the table account_account and then, you use it to refer to account_account_consol_rel. Is there a reason for this ?

veloutin replied:
No, there is no good reason for this and should be made clearer.

dufresnedavid commented on addons/account/account.py:
```diff
@@ -269,15 +269,33 @@ def search(self, cr, uid, args, offset=0, limit=None, order=None,
+                                 WHERE a.child_id = ac.id)
+                )
+                SELECT DISTINCT id, depth
+                     , max(depth) OVER (PARTITION BY id) as max_depth
```
You chose to put the "," after the line break. Is there a particular reason ?

veloutin replied:
I usually format my queries like this, when selecting many fields:

```
SELECT foo
            , bar
            , baz
FROM spam
```

because then I make less errors with trailing commas, especially when reordering fields. Just personal style.

dufresnedavid replied:
Ok thanks, makes sence

StefanRijnhart commented on addons/account/account.py:
```diff
@@ -269,15 +269,33 @@ def search(self, cr, uid, args, offset=0, limit=None, order=None,
+            """,
+            (tuple(ids),)
+        )
+        return [r[0] for r in cr.fetchall()]
```
If you return self.search(cr, uid, ('id', 'in', [r[0] for r in cr.fetchall()], context=context) then you don't loose ORM specific functionality such as the active check and access policy.

veloutin replied:
@StefanRijnhart that does make sense... do you know what the risk is regarding the computation of account balances? I wouldn't want to introduce KeyErrors when a child account is filtered out for some reason before the computation of the account balance.

StefanRijnhart replied:
Well, with my suggestion you apply the same logic as the method you seek to replace, so it should be harmless. In practical terms, journal accounts can be made inactive I believe, so you don't want old configurations with inactive accounts to show up in your reporting here.

Commit:
Optimize recursive account CTE + keep ORM funcs

  Avoid the subselect on account_account_consol_rel which causes
  a whole lot of sequential scans. Instead, use another CTE to
  allow performing a union between the direct parents and consolidated
  parents.

  Also return self.search on returned ids to avoid losing ORM specific
  functionality such as active check and access policy.

Commit:
write a correct search domain

veloutin commented:
For anyone curious about the last change in the query, here are EXPLAIN ANALYZE results for each. This is for getting the children of a single account in a 2000 account table, with 300 entries in the consolidated account relation table. The account tree has 7 levels from the requested account.

**Before**:

```
                                                                        QUERY PLAN                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------
 Subquery Scan on children  (cost=915649.39..915804.06 rows=52 width=12) (actual time=19988.511..19988.616 rows=309 loops=1)
   Filter: (children.depth = children.max_depth)
   ->  Sort  (cost=915649.39..915675.17 rows=10311 width=8) (actual time=19988.506..19988.534 rows=309 loops=1)
         Sort Key: account_children.depth
         Sort Method: quicksort  Memory: 39kB
         CTE account_children
           ->  Recursive Union  (cost=0.00..900223.85 rows=103101 width=8) (actual time=0.113..19987.436 rows=309 loops=1)
                 ->  Index Scan using account_account_pkey on account_account  (cost=0.00..8.33 rows=1 width=4) (actual time=0.103..0.105 rows=1 loops=1)
                       Index Cond: (id = 80)
                 ->  Nested Loop  (cost=0.00..89815.35 rows=10310 width=8) (actual time=1342.303..2855.294 rows=44 loops=7)
                       Join Filter: ((a.parent_id = ac.id) OR (SubPlan 1))
                       ->  WorkTable Scan on account_children ac  (cost=0.00..0.20 rows=10 width=8) (actual time=0.000..0.008 rows=44 loops=7)
                       ->  Materialize  (cost=0.00..6599.87 rows=2058 width=8) (actual time=0.000..0.302 rows=2058 loops=309)
                             ->  Seq Scan on account_account a  (cost=0.00..6589.58 rows=2058 width=8) (actual time=0.027..15.200 rows=2058 loops=1)
                       SubPlan 1
                         ->  Seq Scan on account_account_consol_rel a  (cost=0.00..8.04 rows=5 width=4) (actual time=0.031..0.031 rows=0 loops=635614)
                               Filter: (child_id = ac.id)
         ->  HashAggregate  (cost=14635.11..14738.22 rows=10311 width=8) (actual time=19988.245..19988.332 rows=309 loops=1)
               ->  WindowAgg  (cost=12057.59..13861.85 rows=103101 width=8) (actual time=19987.762..19988.061 rows=309 loops=1)
                     ->  Sort  (cost=12057.59..12315.34 rows=103101 width=8) (actual time=19987.753..19987.806 rows=309 loops=1)
                           Sort Key: account_children.id
                           Sort Method: quicksort  Memory: 39kB
                           ->  CTE Scan on account_children  (cost=0.00..2062.02 rows=103101 width=8) (actual time=0.115..19987.578 rows=309 loops=1)
 Total runtime: 19990.768 ms
(24 rows)
```

It is important to notice that this runs 635k loops on the sequential scan for the consolidated children table. Yikes!

**After**:

```
                                                                             QUERY PLAN                                                                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Subquery Scan on children  (cost=873.71..876.71 rows=1 width=12) (actual time=15.976..16.070 rows=309 loops=1)
   Filter: (children.depth = children.max_depth)
   ->  Sort  (cost=873.71..874.21 rows=200 width=8) (actual time=15.975..16.000 rows=309 loops=1)
         Sort Key: account_children.depth
         Sort Method: quicksort  Memory: 39kB
         CTE account_children
           ->  Recursive Union  (cost=0.00..783.35 rows=861 width=8) (actual time=0.025..15.052 rows=309 loops=1)
                 ->  Index Scan using account_account_pkey on account_account  (cost=0.00..8.27 rows=1 width=4) (actual time=0.019..0.021 rows=1 loops=1)
                       Index Cond: (id = 80)
                 ->  HashAggregate  (cost=74.93..75.79 rows=86 width=8) (actual time=2.073..2.090 rows=44 loops=7)
                       CTE account_children
                         ->  WorkTable Scan on account_children ac  (cost=0.00..0.20 rows=10 width=8) (actual time=0.001..0.010 rows=44 loops=7)
                       ->  Append  (cost=0.33..74.30 rows=86 width=8) (actual time=0.545..2.035 rows=44 loops=7)
                             ->  Hash Join  (cost=0.33..66.12 rows=40 width=8) (actual time=0.531..1.794 rows=44 loops=7)
                                   Hash Cond: (aa.parent_id = ac.id)
                                   ->  Seq Scan on account_account aa  (cost=0.00..57.58 rows=2058 width=8) (actual time=0.004..0.530 rows=2058 loops=7)
                                   ->  Hash  (cost=0.20..0.20 rows=10 width=8) (actual time=0.062..0.062 rows=44 loops=7)
                                         Buckets: 1024  Batches: 1  Memory Usage: 1kB
                                         ->  CTE Scan on account_children ac  (cost=0.00..0.20 rows=10 width=8) (actual time=0.002..0.038 rows=44 loops=7)
                             ->  Hash Join  (cost=0.33..7.31 rows=46 width=8) (actual time=0.215..0.215 rows=0 loops=7)
                                   Hash Cond: (aa_rel.child_id = ac.id)
                                   ->  Seq Scan on account_account_consol_rel aa_rel  (cost=0.00..5.21 rows=321 width=8) (actual time=0.008..0.083 rows=321 loops=7)
                                   ->  Hash  (cost=0.20..0.20 rows=10 width=8) (actual time=0.040..0.040 rows=44 loops=7)
                                         Buckets: 1024  Batches: 1  Memory Usage: 1kB
                                         ->  CTE Scan on account_children ac  (cost=0.00..0.20 rows=10 width=8) (actual time=0.001..0.016 rows=44 loops=7)
         ->  HashAggregate  (cost=80.72..82.72 rows=200 width=8) (actual time=15.857..15.902 rows=309 loops=1)
               ->  WindowAgg  (cost=59.19..74.26 rows=861 width=8) (actual time=15.473..15.730 rows=309 loops=1)
                     ->  Sort  (cost=59.19..61.35 rows=861 width=8) (actual time=15.469..15.503 rows=309 loops=1)
                           Sort Key: account_children.id
                           Sort Method: quicksort  Memory: 39kB
                           ->  CTE Scan on account_children  (cost=0.00..17.22 rows=861 width=8) (actual time=0.027..15.296 rows=309 loops=1)
 Total runtime: 16.246 ms
```

Commit:
remove parents around list of ids

StefanRijnhart commented:
Great work! Please file the upstream PR and we should be good, right?

veloutin commented:
Upstream: https://github.com/odoo/odoo/pull/7567

Not sure how likely to be accepted though.

bwrsandman commented:
You could probably make the title more sexy like `[OPTIMIZATION] 1000x faster on account children call` ;)

veloutin commented:
I suppose I could test how fast it is with the recursive python version first! :P

agb80 commented:
:+1:

veloutin commented:
I have run a sample test between the original python version and this version. For the id that was causing me troubles with the slow query, the SQL version was about 5-10% faster than the python version.

ghost commented:
:+1: @veloutin Very nice work

gurneyalex commented:
:+1:
